pycalphad/core/eqsolver.pyx

```diff
@@ -82,6 +81,50 @@ cdef bint add_new_phases(object composition_sets, object removed_compsets, objec
         return True
     return False
 
+@cython.boundscheck(False)
+cdef int argmax(double* a, int a_shape) nogil:
+    cdef int i
+    cdef int result = 0
+    cdef double highest = -1e30
+    for i in range(a_shape):
+        if a[i] > highest:
+            highest = a[i]
+            result = i
+    return result
+
+def add_nearly_stable(object composition_sets, dict phase_records,
+                      object grid, object current_idx, np.ndarray[ndim=1, dtype=np.float64_t] chemical_potentials,
+                      double[::1] state_variables, double minimum_df, bint verbose):
+    cdef double[::1] driving_forces, driving_forces_for_phase
+    cdef double[:,::1] current_grid_Y = grid.Y[*current_idx, ...]
+    cdef double[:,::1] current_grid_X = grid.X[*current_idx, ...]
+    cdef double[::1] current_grid_GM = grid.GM[*current_idx, ...]
+    cdef unicode phase_name
+    cdef CompositionSet compset = composition_sets[0]
+    cdef set stable_phases = {compset.phase_record.phase_name for compset in composition_sets}
+    cdef PhaseRecord phase_record
+    cdef int num_statevars = len(compset.phase_record.state_variables)
+    cdef int df_idx, minimum_df_idx
+    cdef bint phases_added = False
+    driving_forces = np.dot(current_grid_X, chemical_potentials) - current_grid_GM
```

Comment:
Does it make more sense to allocate driving forces outside `add_nearly_stable` and pass the array to the `out` keyword argument of `np.dot`? For broadcasted equilibrium calculations, this will allocate an array of the same size as `grid.GM` for every set of conditions, but the size of the grid doesn't change.

Profiling would reveal if it's worth doing.

Reply:
It's only worth it if we call `add_nearly_stable` inside the solve loop, which I'm not sure that we should at this time.